### DISCUSSION TITLE
Optionally increase annotation tag column length

### DISCUSSION THREAD

**Discussion by papagian** [grafana/grafana#51815] [Platform]
Follow up of the grafana/support-escalations#3134
[internal discussion](https://raintank-corp.slack.com/archives/C02DJ3HV42W/p1657009218685979)
Currently annotation update fails if the tags exceed that maximum column length (500).

Suggested improvement:
* change the MySQL column type to `TEXT` or `MEDIUMTEXT`(so that no limit is enforced on the database level).
* enforce a configurable limit (defaulting to `500`) in the code.
* warn users that values beyond the default aren’t guaranteed to perform well

There is an additional suggestion (to consider) for introducing a new column (with no database limit) instead of altering the existing one.

**Comment by dafydd-t** [comment]
Some opinions from me:\

- Allowing a runtime config item to change the schema seems unexpected. Its likely to introduce complexity, such as: will we allow this config value to be decreased? We'd have to add checks.
- If we're considering changing the column type to `TEXT|MEDIUMTEXT`, why not just change the column type to `VARCHAR(65535)` (the max), which is equivalent to the max for `TEXT` (not medium text). This would not require expand-contract as we'd "just" be increasing the limit on an existing column. ( think length 65535 is enough for any sane use, vs. moving to the higher length limit provided by `MEDIUMTEXT`.
- I don't think there is much storage impact on using `VARCHAR` vs `TEXT`: `VARCHAR` still only uses 2 + `c` bytes of disk where `c` is the length of the string (when `c` > 255), which is the same as `TEXT`.

**Reply by papagian** [reply]
> Some opinions from me:\
> 
> * Allowing a runtime config item to change the schema seems unexpected. Its likely to introduce complexity, such as: will we allow this config value to be decreased? We'd have to add checks.

Sorry for the confusion; what I suggest is to prevent to store annotations tags larger than this configuration, not to change the table schema based on this configuration.

> * If we're considering changing the column type to `TEXT|MEDIUMTEXT`, why not just change the column type to `VARCHAR(65535)` (the max), which is equivalent to the max for `TEXT` (not medium text). This would not require expand-contract as we'd "just" be increasing the limit on an existing column. ( think length 65535 is enough for any sane use, vs. moving to the higher length limit provided by `MEDIUMTEXT`.

Since this column is not part of an index, its type does not make big difference so no objections. However, given that SQLite [does not support `ALTER COLUMN`](https://www.sqlite.org/omitted.html) ~we still have to drop and re-create the column for increasing the column size (for at least this driver).~ SQLite [does not enforce the length of a VARCHAR](https://www.sqlite.org/faq.html#q9) so we only need to change it for MySQL and PostgresSQL.

**Reply by dafydd-t** [reply]
> Sorry for the confusion; what I suggest is to prevent to store annotations tags larger than this configuration, not to change the table schema based on this configuration.

Great!

**Comment by papagian** [comment]
Regarding `VARCHAR` vs`TEXT|MEDIUMTEXT`, there is also [maximum row size limit](https://dev.mysql.com/doc/refman/8.0/en/column-count-limit.html#row-size-limits) of `65,535` bytes in MySQL.
`VARCHAR` type contributes to it while `BLOB` and `TEXT` columns only contribute 9 to 12 bytes because their contents are stored separately from the rest of the row. Depending on the rest column types of the table, in case of its violation [ER_TOO_BIG_FIELDLENGTH](https://dev.mysql.com/doc/mysql-errors/8.0/en/server-error-reference.html#error_er_too_big_fieldlength) or [ER_TOO_BIG_ROWSIZE](https://dev.mysql.com/doc/mysql-errors/8.0/en/server-error-reference.html#error_er_too_big_rowsize) can occur.
Therefore, we either have to compute and set the maximum possible length for the `tags` column (it can be problematic if more columns will be added in the future) or switch to `TEXT|MEDIUMTEXT`

**Reply by dafydd-t** [reply]
interesting, thanks for sharing. I have no concerns about either approach then - I just ask to make sure that any type changes are backwards compatible, and if they're not, then to use [expand-contract pattern](https://medium.com/continuousdelivery/expand-contract-pattern-and-continuous-delivery-of-databases-4cfa00c23d2e) to make sure we don't cause disruption for HG instances during upgrade.

**Reply by papagian** [reply]
> interesting, thanks for sharing. I have no concerns about either approach then - I just ask to make sure that any type changes are backwards compatible, and if they're not, then to use [expand-contract pattern](https://medium.com/continuousdelivery/expand-contract-pattern-and-continuous-delivery-of-databases-4cfa00c23d2e) to make sure we don't cause disruption for HG instances during upgrade.

I have decided in favour of the `VARCHAR` to make the migration less painful.

**Comment by papagian** [comment]
The PR: #54754

**Comment by tolzhabayev** [comment]
Seems like this discussion can be locked then? @papagian

**Reply by papagian** [reply]
yes!

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
